controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,466 @@
+		if k8serrors.IsNotFound(err) {
+			// A resource doesn't exist, create one
+			if err = r.newHostFirmwareSettings(info, req.NamespacedName); err != nil {
+
+				return ctrl.Result{}, errors.Wrap(err, "could not create firmware settings")
+			}
+		} else {
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+		}
+	}
+
+	// Create a provisioner that can access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	ready, err := prov.IsReady()
+	if err != nil || !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	// Get the data from Ironic and update HFS, the call to provisioner may fail if settings can't
+	// be retrieved for the bios_interface.
+	if err = r.updateHostFirmwareSettings(prov, info); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Could not update hostFirmwareSettings")
+	}
+
+	for _, e := range info.events {
+		r.publishEvent(req, e)
+	}
+
+	// only run again on changes
+	return ctrl.Result{}, nil
+}
+
+// Get the firmware settings from the provisioner and update hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) updateHostFirmwareSettings(prov provisioner.Provisioner, info *rInfo) (err error) {
+
+	info.log.Info("retrieving firmware settings and saving to resource", "node", info.bmh.Status.Provisioning.ID)
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		return errors.Wrap(err, "could not get firmware settings from provisioner")
+	}
+
+	// get or create a firmwareSchema to hold schema
+	firmwareSchema, err := r.getOrCreateFirmwareSchema(info, schema)
+	if err != nil {
+		return errors.Wrap(err, "could not get/create firmware schema")
+	}
+
+	// Set hostFirmwareSetting to use this schema
+	info.hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+		Namespace: firmwareSchema.ObjectMeta.Namespace,
+		Name:      firmwareSchema.ObjectMeta.Name}
+
+	if err = r.updateStatus(info, currentSettings, firmwareSchema); err != nil {
+		return errors.Wrap(err, "could not update hostFirmwareSettings")
+	}
+
+	return nil
+}
+
+// Update the HostFirmwareSettings resource using the settings and schema from provisioner
+func (r *HostFirmwareSettingsReconciler) updateStatus(info *rInfo, settings metal3v1alpha1.SettingsMap, schema *metal3v1alpha1.FirmwareSchema) (err error) {
+
+	if info.hfs.Status.Settings == nil {
+		info.hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	}
+
+	// Update Status
+	for k, v := range settings {
+		// Some vendors include encrypted password fields, don't add these
+		if strings.Contains(k, "Password") {
```

Comment:
OK, I'm recalling some stuff now and there's a bit of a usability issue here. Someone trying to clone the settings from one host to another can't just copy the settings from a Status to a Spec because the Status includes read-only settings.
In the design doc we said we were going to solve this by copying all of the writeable values from the Status to the Spec, but we decided not to do that.
Maybe the Status needs to be split into "Settings" (writeable) and "Data" (read-only). WDYT?

Reply:
> LocalSettings?
> 
I like that

> Btw, IIRC this part will also be addressed in future by another controller (to copy the right settings to other hosts to keep them in sync)
Yes that is the plan, although from the VRAN discussions the importance was more on configuring the settings on a host during initial deployment than on keeping them in sync post-deployment